src/pykeen/triples/analysis.py

```diff
@@ -0,0 +1,417 @@
+"""Analysis utilities for (mapped) triples."""
+import hashlib
+import itertools as itt
+import logging
+from collections import defaultdict
+from typing import Collection, DefaultDict, Iterable, Mapping, NamedTuple, Set, Tuple
+
+import numpy
+import pandas as pd
+import torch
+from tqdm.auto import tqdm
+
+logger = logging.getLogger(__name__)
+
+__all__ = [
+    "relation_cardinality_types",
+    "relation_pattern_types",
+]
+
+# constants
+CARDINALITY_TYPE_ONE_TO_ONE = "one-to-one"
+CARDINALITY_TYPE_ONE_TO_MANY = "one-to-many"
+CARDINALITY_TYPE_MANY_TO_ONE = "many-to-one"
+CARDINALITY_TYPE_MANY_TO_MANY = "many-to-many"
+
+relation_cardinalities_types = {
```

Comment:
somehow an enum would be appropriate here, but I'm not a big fan of how they work. so this is sad.

Reply:
tried that, made some accessibility problems with having them in dataframes, and thus reverted again.